### PR TITLE
feat(api,hooks): add v2 hook signature and github repository analysis

### PR DESCRIPTION
**Pull request by sguiheux**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

File: engine/api/v2_project.go [hunk 1]
```diff
@@ -20,3 +25,25 @@ func (api *API) getAllRepositoriesHandler() ([]service.RbacChecker, service.Hand
...
 		}
 }
+
+func (api *API) getRepositoryHookHandler() ([]service.RbacChecker, service.Handler) {
+	return service.RBAC(rbac.IsHookService),
+		func(ctx context.Context, w http.ResponseWriter, req *http.Request) error {
+			vars := mux.Vars(req)
+			repoIdentifier, err := url.PathUnescape(vars["repositoryIdentifier"])
+			if !sdk.IsValidUUID(repoIdentifier) {
+				return sdk.NewErrorFrom(sdk.ErrWrongRequest, "this handler need the repository uuid")
```
Comment (yesnault) [inline engine/api/v2_project.go:35]:
```suggestion
				return sdk.NewErrorFrom(sdk.ErrWrongRequest, "this handler needs the repository uuid")
```

**Review by yesnault — APPROVED** [review]

File: engine/hooks/hooks_router.go [hunk 2]
```diff
@@ -48,6 +57,71 @@ func (s *Service) initRouter(ctx context.Context) {
...
+		h := hmac.New(sha256.New, []byte(hook.HookSignKey))
+		h.Write(body)
+		sha := hex.EncodeToString(h.Sum(nil))
+
+		if strings.TrimPrefix(signHeaderValue, "sha256=") != sha {
+			log.Error(ctx, "signature mismatch: got %s, compute %s", signHeaderValue, sha)
+			return ctx, sdk.NewErrorFrom(sdk.ErrUnauthorized, "wrong signature")
+		}
+
+		ctx = context.WithValue(ctx, "body", body)
```
Comment (fsamin) [inline engine/hooks/hooks_router.go:120]: don't put the body in the context, you have to reset a readable body in the request itself

File: engine/hooks/hooks_router.go [hunk 3]
```diff
@@ -48,6 +57,71 @@ func (s *Service) initRouter(ctx context.Context) {
...
+		if tokenHeaderValue != hook.HookSignKey {
+			return ctx, sdk.NewErrorFrom(sdk.ErrUnauthorized, "token mismatch")
+		}
+
+		body, err := io.ReadAll(req.Body)
+		if err != nil {
+			return ctx, sdk.NewErrorFrom(sdk.ErrUnauthorized, "unable to check signature")
+		}
+
+		ctx = context.WithValue(ctx, "body", body)
```
Comment (fsamin) [inline engine/hooks/hooks_router.go:85]: same as below

File: engine/hooks/hooks_handlers.go [hunk 4]
```diff
@@ -126,10 +126,14 @@ func (s *Service) repositoryWebHookHandler() service.Handler {
 		vars := mux.Vars(r)
 		uuid := vars["uuid"]
 
-		defer r.Body.Close()
-		body, err := io.ReadAll(r.Body)
-		if err != nil {
-			return sdk.NewErrorFrom(sdk.ErrUnknownError, "unable to read body: %v", err)
+		body, ok := ctx.Value("body").([]byte)
```
Comment (fsamin) [inline engine/hooks/hooks_handlers.go:129]: read the body from the request (the middleware shouldn't mutate the request)

**Review by fsamin — CHANGES_REQUESTED** [review]

File: engine/api/repository/dao_vcs_project_repository.go [hunk 5]
```diff
@@ -102,6 +97,15 @@ func LoadAllRepositoriesByVCSProjectID(ctx context.Context, db gorp.SqlExecutor,
 	return repositories, nil
 }
 
+func LoadRepositoryByID(ctx context.Context, db gorp.SqlExecutor, id string, opts ...gorpmapping.GetOptionFunc) (sdk.ProjectRepository, error) {
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply engine/api/repository/dao_vcs_project_repository.go:100]: done

File: engine/api/v2_project_repository.go [hunk 6]
```diff
@@ -206,3 +213,41 @@ func (api *API) getVCSProjectRepositoryAllHandler() ([]service.RbacChecker, serv
...
+			srvs, err := services.LoadAllByType(ctx, api.mustDB(), sdk.TypeHooks)
+			if err != nil {
+				return err
+			}
+			if len(srvs) == 0 {
+				return sdk.NewErrorFrom(sdk.ErrNotFound, "no hook service found")
+			}
+			hook := sdk.HookAccessData{
+				URL:         fmt.Sprintf("%s/v2/webhook/repository/%s/%s", srvs[0].HTTPURL, vcsProject.Type, repo.ID),
+				HookSignKey: repo.HookSignKey,
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply engine/api/v2_project_repository.go:249]: done

File: engine/hooks/hooks_router.go [hunk 7]
```diff
@@ -48,6 +57,71 @@ func (s *Service) initRouter(ctx context.Context) {
...
+		}
+
+		defer req.Body.Close()
+		body, err := io.ReadAll(req.Body)
+		if err != nil {
+			return ctx, sdk.NewErrorFrom(sdk.ErrUnauthorized, "unable to check signature")
+		}
+
+		// Create a new HMAC by defining the hash type and the key (as byte array)
+		h := hmac.New(sha256.New, []byte(hook.HookSignKey))
```
[the remark this answers is not in the document]
Comment (sguiheux) [reply engine/hooks/hooks_router.go:111]: 5ms on a gitea event

File: engine/hooks/hooks_router.go (hunk 3 above)
Comment (sguiheux) [reply engine/hooks/hooks_router.go:85]: done

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#17801.0 ✘  
* Build
  * Build and Package All ✘  
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔

File: engine/api/operation/operation.go [hunk 8]
```diff
@@ -151,6 +152,8 @@ func GetRepositoryOperation(ctx context.Context, db gorp.SqlExecutor, uuid strin
 
 // Poll repository operation for given uuid.
 func Poll(ctx context.Context, db gorp.SqlExecutor, operationUUID string) (*sdk.Operation, error) {
+	_, next := telemetry.Span(ctx, "operation.Poll")
```
Comment (fsamin) [inline engine/api/operation/operation.go:155]: you should keep the context to use it later

File: engine/api/rbac/dao_rbac_project_key.go [hunk 9]
```diff
@@ -51,6 +52,8 @@ func loadRRBACProjectKeys(ctx context.Context, db gorp.SqlExecutor, rbacProjectI
 }
 
 func HasRoleOnProjectAndUserID(ctx context.Context, db gorp.SqlExecutor, role string, userID string, projectKey string) (bool, error) {
+	_, next := telemetry.Span(ctx, "rbac.HasRoleOnProjectAndUserID")
```
Comment (fsamin) [inline engine/api/rbac/dao_rbac_project_key.go:55]: keep the context

File: engine/api/repository/dao_vcs_project_repository.go [hunk 10]
```diff
@@ -102,6 +98,17 @@ func LoadAllRepositoriesByVCSProjectID(ctx context.Context, db gorp.SqlExecutor,
 	return repositories, nil
 }
 
+func LoadRepositoryByID(ctx context.Context, db gorp.SqlExecutor, id string, opts ...gorpmapping.GetOptionFunc) (*sdk.ProjectRepository, error) {
+	_, next := telemetry.Span(ctx, "repository.LoadRepositoryByID")
```
Comment (fsamin) [inline engine/api/repository/dao_vcs_project_repository.go:102]: keep the context

File: engine/api/repository/dao_vcs_project_repository_analyze.go [hunk 11]
```diff
@@ -111,6 +112,8 @@ func LoadRepositoryIDsAnalysisInProgress(ctx context.Context, db gorp.SqlExecuto
 }
 
 func LoadRepositoryAnalysisById(ctx context.Context, db gorp.SqlExecutor, projectRepoID, analysisID string) (*sdk.ProjectRepositoryAnalysis, error) {
+	_, next := telemetry.Span(ctx, "repository.LoadRepositoryAnalysisById")
```
Comment (fsamin) [inline engine/api/repository/dao_vcs_project_repository_analyze.go:115]: keep the context

File: engine/api/user/dao_gpgkey.go [hunk 12]
```diff
@@ -64,6 +65,8 @@ func LoadGPGKeysByUserID(ctx context.Context, db gorp.SqlExecutor, userID string
 }
 
 func LoadGPGKeyByKeyID(ctx context.Context, db gorp.SqlExecutor, keyID string) (*sdk.UserGPGKey, error) {
+	_, next := telemetry.Span(ctx, "user.LoadGPGKeyByKeyID")
```
Comment (fsamin) [inline engine/api/user/dao_gpgkey.go:68]: keep the context

File: engine/api/v2_repository_analyze.go [hunk 13]
```diff
@@ -294,171 +233,256 @@ func (api *API) repositoryAnalysisPoller(ctx context.Context, tick time.Duration
...
+	}
+	if ope.Status == sdk.OperationStatusError {
+		analysis.Data.Error = ope.Error.Message
+		analysis.Status = sdk.RepositoryAnalysisStatusError
+	}
+	return nil
 }
 
 func (api *API) getCdsFilesOnVCSDirectory(ctx context.Context, client sdk.VCSAuthorizedClientService, analysis *sdk.ProjectRepositoryAnalysis, repoName, commit, directory string) error {
+	_, next := telemetry.Span(ctx, "api.getCdsFilesOnVCSDirectory")
```
Comment (fsamin) [inline engine/api/v2_repository_analyze.go:484]: keep the context

File: engine/api/v2_repository_analyze.go [hunk 14]
```diff
@@ -480,6 +504,8 @@ func (api *API) getCdsFilesOnVCSDirectory(ctx context.Context, client sdk.VCSAut
 }
 
 func (api *API) getCdsArchiveFileOnRepo(ctx context.Context, client sdk.VCSAuthorizedClientService, repo sdk.ProjectRepository, analysis *sdk.ProjectRepositoryAnalysis) error {
+	_, next := telemetry.Span(ctx, "api.getCdsArchiveFileOnRepo")
```
Comment (fsamin) [inline engine/api/v2_repository_analyze.go:507]: keep the context

File: engine/api/vcs/dao_vcs_project.go [hunk 15]
```diff
@@ -84,6 +85,8 @@ func LoadVCSByProject(ctx context.Context, db gorp.SqlExecutor, projectKey strin
 }
 
 func LoadVCSByID(ctx context.Context, db gorp.SqlExecutor, projectKey string, vcsID string, opts ...gorpmapping.GetOptionFunc) (*sdk.VCSProject, error) {
+	_, next := telemetry.Span(ctx, "vcs.LoadVCSByID")
```
Comment (fsamin) [inline engine/api/vcs/dao_vcs_project.go:88]: keep the context

**Review by fsamin — CHANGES_REQUESTED** [review]

**Review by fsamin — APPROVED** [review]

**Review by richardlt — APPROVED** [review]

**Pull request merged**